File: Sharebook/Client/src/app/core/Core/Helpers/IJsonObjectConverter.ts
import {inject, Inject, InjectionToken} from "@angular/core";

import {JsonObjectConverter} from "../JsonConverters/JsonObjectConverter";
import {JsonConvertWrapper} from "../../../logic/Wrapper/JsonConvertWrapper";

// This interface allows to (de)serialize any object or interface from the Tweetinvi API
export interface IJsonObjectConverter {
  Serialize(o: object, converters?: any[]): string;

  Deserialize<T>(json: string, converters?: any[]): T;
}

export const IJsonObjectConverterToken = new InjectionToken<IJsonObjectConverter>('IJsonObjectConverter', {
  providedIn: 'root',
  factory: () => new JsonObjectConverter(inject(JsonConvertWrapper)),
});

File: Server/src/Chessbook.Core/JsonConverters/IJsonObjectConverter.cs
using Newtonsoft.Json;

namespace Chessbook.Core.JsonConverters
{
    /// <summary>
    /// This interface allows to (de)serialize any object or interface from the Tweetinvi API
    /// </summary>
    public interface IJsonObjectConverter
    {
        string Serialize(object o, JsonConverter[] converters = null);
        T Deserialize<T>(string json, JsonConverter[] converters = null);
    }
}
